ports/rp2: Added init and reset C hooks for RP2040, and increased external pin count

ZodiusInfuser opened the pull request:
Following discussions in https://github.com/orgs/micropython/discussions/12116, this pull request makes changes primarily to support the upcoming Pimoroni Yukon board, but that could also benefit others in the community.

* Adds C hooks for initialising and resetting an RP2040 based board. Yukon currently only uses a few of these, but I included the rest for future product use.
* Increases the number of external IO pins from 10 to the maximum 32 to match the number of pins Yukon's IO expanders expose.

I have tested these successfully with a custom build, that includes the board definitions and pulls in drivers for the IO expanders so the external pins can work. I have left these changes out of this PR, as they are currently quite intrusive.

Gadgetoid commented:
Could there be a means to include a C header, for hook code that requires it? Or some knack to it?

In my case - for PicoSystem - I might want `MICROPY_BOARD_STARTUP` to be:

```c
// Apply a modest overvolt, default is 1.10v.
// this is required for a stable 250MHz on some RP2040s
vreg_set_voltage(VREG_VOLTAGE_1_20);
sleep_ms(10);
set_sys_clock_khz(250000, true);
```

But that depends upon:

```c
#include "hardware/vreg.h
```

But yeah, +1 to this change, it could benefit PicoSystem and potentially PicoVision.

ZodiusInfuser commented:
I think that may already be covered by this addition, as I do

```c
#include "hardware/gpio.h"
#include "hardware/i2c.h"
```
within my Yukon build. See: https://github.com/pimoroni/yukon/blob/main/firmware/PIMORONI_YUKON/board.c

ZodiusInfuser commented:
Hi. Just wondering if anyone had a chance to look at this yet?

projectgus commented on ports/rp2/main.c:
```diff
@@ -199,6 +208,10 @@ int main(int argc, char **argv) {
         mp_printf(MP_PYTHON_PRINTER, "MPY: soft reboot\n");
+
+        // Hook for resetting anything immediately following a soft reboot command
+        MICROPY_BOARD_EARLY_RESET();
```
As above, a similar hook in the stm32 and renesas-ra ports is `MICROPY_BOARD_START_SOFT_RESET(&state)`. Although these hooks have default values on these ports (and are in a slightly different position), would it be possible to make this hook the same as stm32/renesas for consistency?

ZodiusInfuser replied:
This one I actively use, but I am happy to change it.

The different line seems to be because it is called before the `mp_printf` on STM, which I think is the better approach, as that state also gives the option to disable the print if a build requires.

projectgus commented on ports/rp2/main.c:
```diff
@@ -213,6 +226,9 @@ int main(int argc, char **argv) {
         mp_deinit();
+
+        // Hook for resetting anything after the MicroPy runtime has been deinitialised
+        MICROPY_BOARD_LATE_RESET();
```
One more of the same. The equivalent hook for stm32 and renesas-ra is `MICROPY_BOARD_END_SOFT_RESET(&state)`, and it's placed before `gc_sweep_all()`. Would that work for your use case, also?

ZodiusInfuser replied:
I do not actively use this hook, so am happy to move it to before gc_sweep_all().

Though, I wonder if there are use-cases where you would want a call to occur post `mp_deinit()` 🤔. I assume your team chose its location before the gc for a reason.

projectgus replied:
> Though, I wonder if there are use-cases where you would want a call to occur post mp_deinit() 🤔. I assume your team chose its location before the gc for a reason.

Good questions... I don't know the historical reason these hooks were added (there is some discussion in #6505 but it doesn't cover that.)

That said, I believe the reason to do it before `gc_sweep_all()` is to give you a chance to NULL out any pointers into the Python heap. If you have, for example, a C based interrupt handler that updates some data in the Python heap then that needs to be disabled and the pointer NULLed out before `gc_sweep_all()` blows away all of the Python heap contents. Otherwise it's going to be left with a dangling pointer to uninitialized Python heap space, and there may be a race condition where an interrupt fires and accesses that memory.

> I do not actively use this hook, so am happy to move it to before gc_sweep_all().

Sounds good, thanks. I'm wondering about the relative benefits of trimming back the hooks in this PR to just the ones you are using. Although I'm leaning towards the best outcome being if the hooks in rp2 match the ones in stm32/renesas, so it's worth adding them all if it achieves that goal.

projectgus commented on ports/rp2/main.c:
```diff
@@ -166,6 +172,9 @@ int main(int argc, char **argv) {
         #endif
 
+        // Hook for setting up anything that should follow the MicroPy runtime but be prior to _boot.py
+        MICROPY_BOARD_LATE_INIT();
```
From a maintainability perspective, I have some concerns about having different board-level hooks on different ports that implement similar or the same functionality.

The way this is done on the stm32 & renesas-ra ports is there is a hook `MICROPY_BOARD_RUN_BOOT_PY(&state)` which calls `boardctrl_run_boot_py()` by default, but can be hooked to do something else instead.

I'm not sure that's necessarily a better design, but I am sure that doing it differently per-port will eventually become a pain to maintain.

How would you feel about refactoring the boot.py logic immediately below this into a separate function `rp2_run_boot_py` or similar, and have his hook call there by default or optionally be overridden?

ZodiusInfuser replied:
As mentioned in my main reply, I'm not actively using this hook, so it could just be removed for the purpose of getting this PR through.

That being said, I do patch the Pimoroni MicroPython build for Yukon with a similar `state` for determining if boot and main py should run, and what you propose would let me do this in an official way without the need to keep a fork active, if I am understanding correctly.

```c
    if (run_main) {
            // Execute user scripts.
            int ret = pyexec_file_if_exists("boot.py");
            if (ret & PYEXEC_FORCED_EXIT) {
                run_main = !(ret & PYEXEC_SKIP_USER_CODE);
                goto soft_reset_exit;
            }
            if (pyexec_mode_kind == PYEXEC_MODE_FRIENDLY_REPL) {
                ret = pyexec_file_if_exists("main.py");
                if (ret & PYEXEC_FORCED_EXIT) {
                    run_main = !(ret & PYEXEC_SKIP_USER_CODE);
                    goto soft_reset_exit;
                }
            }
        }
        run_main = true;
```

My only hesitation with me doing what you propose (versus someone more familiar with the codebase) is the bigger impact it may have on other RP2 products, since I am likely to break things.

projectgus replied:
> what you propose would let me do this in an official way without the need to keep a fork active, if I am understanding correctly.

I think so, yes.

> is the bigger impact it may have on other RP2 products, since I am likely to break things.

I think the chance of breakage is pretty low, provided the default implementation of the hook calls a function that perform the "Execute user scripts" stage as it's implemented now. In that case the only change is really the additional function call and maybe a return check. I'm very happy to review such a change.

For a reference of the "pattern" for such an implementation, the stm32 port is a good reference:
https://github.com/micropython/micropython/blob/master/ports/stm32/main.c#L578
https://github.com/micropython/micropython/blob/master/ports/stm32/boardctrl.h#L64
https://github.com/micropython/micropython/blob/master/ports/stm32/boardctrl.c#L192

This raises the question of whether rp2 port should add its own boardctrl.[hc] as well. Thinking about it now, I feel this might be a good refactoring to keep these new features well-defined, even if it's a pretty minimal implementation. However I'm open to other approaches, too. :)

All that said, if you don't actually need this right now then pulling it out of this PR and potentially doing it in a future PR is also an option.

ZodiusInfuser replied:
Okay. I've updated this PR to only have 4 hooks, and renamed them to match the STM32's naming scheme.

Although it would be nice to have the boot.py and main.py stages follow the pattern used by STM32, at this time I do not need that level of control for Yukon. Turns out that code I showed that modified the main run process never actually did anything different as the `PYEXEC_SKIP_USER_CODE` value I added was never actually emitted by anything.

projectgus reviewed (changes requested):
Hi @ZodiusInfuser,

Thanks for being patient while I got around to looking at this.

I think these are useful changes, and thanks for contributing them.

The max EXT_PINs thing is good to go (I was initially cautious until I realised this only raises the maximum to 32, so doesn't change anything on the current boards!)

For the new C hooks, these are great to have on rp2. My only concern is consistency between ports. See the inline comments for specific thoughts.

ZodiusInfuser commented:
HI @projectgus,

Thank you for taking a look at this.

> The max EXT_PINs thing is good to go (I was initially cautious until I realised this only raises the maximum to 32, so doesn't change anything on the current boards!)

Glad to hear it!
 
> For the new C hooks, these are great to have on rp2. My only concern is consistency between ports. See the inline comments for specific thoughts.

I hadn't specifically spotted the STM etc hooks, so apologies for their inconsistency. I will have a look and see what can be changed and respond to your specific notes later.

I will say, I am only actively using the below hooks in the product this PR was primarily raised for.

```c
#define MICROPY_BOARD_EARLY_INIT board_init
#define MICROPY_BOARD_EARLY_RESET board_reset
``` 

The others were included to account for possible future use-cases, so I agree they should be consistent.

projectgus commented:
@ZodiusInfuser Are you still working on this PR? Is there anything you're waiting on from the MicroPython side?

ZodiusInfuser commented:
Errr... I'll admit I completely forgot about this open PR after our Pimoroni Yukon was launched 😅. It is still necessary as that product's builds still rely on my `rp_init_hooks` branch.

Looks like I never addressed your review points before. I'll try and find time this week to look at it, if not it will be after I am back from holiday.

----------
Also tangentially, I noticed in `machine_pin_cyw43.c` that the opening check assumes that if the network is defined and external pins are set, then those pins come from the cyw43. As micropython does not yet support multiple external pin types (unless things have moved on in the last 9 months?), I believe the following line:
```c
#if defined(MICROPY_PY_NETWORK_CYW43) && defined(MICROPY_HW_PIN_EXT_COUNT)
```
should be changed to:
```c
#if CYW43_GPIO == 1 && defined(MICROPY_PY_NETWORK_CYW43) && defined(MICROPY_HW_PIN_EXT_COUNT)
```
to support any future Pico W-based Yukon I may attempt.

Would that be something I could include in this PR, or raise as an issue / separate PR?

ZodiusInfuser committed:
ports/rp2: Additions for rp2 init hooks.

Signed-off-by: Christopher Parrott <chris@pimoroni.com>

ZodiusInfuser committed:
ports/rp2: Increased ext pins to 32.

Signed-off-by: Christopher Parrott <chris@pimoroni.com>

projectgus reviewed (approved):
This looks good to me, thanks for updating and I hope it makes maintenance easier for the Yukon and other boards!

projectgus commented:
>  Would that be something I could include in this PR, or am I better to raise it as a separate issue / PR?

This `machine_pin_cyw43.c` change looks very reasonable to me, please free to submit it in a separate PR.

dpgeorge commented:
This looks good.  Thanks for making it more consistent with the stm32 port.

dpgeorge commented:
Rebased and merged in 5dfd3ecd8b1c99ae08ce082102606739bde57852 and 79ba6d8ce7f322baa8d606e52672c1bd4b245bb0
